llrt_utils/src/bytes.rs

```diff
@@ -1,8 +1,247 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
+
+use std::borrow::Cow;
+
 use rquickjs::{ArrayBuffer, Coerced, Ctx, Exception, IntoJs, Object, Result, TypedArray, Value};
 
-use super::result::ResultExt;
+#[derive(Clone)]
+pub enum ObjectBytes<'js> {
+    U8Array(TypedArray<'js, u8>),
+    I8Array(TypedArray<'js, i8>),
+    U16Array(TypedArray<'js, u16>),
+    I16Array(TypedArray<'js, i16>),
+    U32Array(TypedArray<'js, u32>),
+    I32Array(TypedArray<'js, i32>),
+    U64Array(TypedArray<'js, u64>),
+    I64Array(TypedArray<'js, i64>),
+    F32Array(TypedArray<'js, f32>),
+    F64Array(TypedArray<'js, f64>),
+    DataView(ArrayBuffer<'js>),
+    Vec(Option<Vec<u8>>),
+}
+
+impl<'js> ObjectBytes<'js> {
+    const DETACHED_ERROR: &'static str = "ArrayBuffer is detached";
+
+    pub fn from(ctx: &Ctx<'js>, value: Value<'js>) -> Result<Self> {
+        Self::from_offset(ctx, value, 0, None)
+    }
+
+    pub fn from_offset(
+        ctx: &Ctx<'js>,
+        value: Value<'js>,
+        offset: usize,
+        length: Option<usize>,
+    ) -> Result<Self> {
+        if value.is_undefined() {
+            return Ok(ObjectBytes::Vec(Some(vec![])));
+        }
+        if let Some(bytes) = get_string_bytes(&value, offset, length)? {
```

Comment:
A lot of the current llrt APIs are not fully compliant with node because we miss the encoding part for strings. If you want to bridge that gap we will need to have another input for that function

Reply:
I think most does. This utility does not handle encodings as it’s a lower level utility just to get the raw bytes. As you can see in the encoder related structs we’re transforming the bytes with respect to their encoding property